### PR TITLE
Feature/buff window

### PR DESCRIPTION
**Pull request by EvgenijK**
It would be nice to have indicators for some timed effects on character. Especially for resistances.
There is even a 4-year old topic on forum about it https://tomenet.eu/phpBB3/viewtopic.php?f=7&t=932

This code adds indicators for timed four basic resistances, poison resistance, mana resistance and timed full ESP.
Indicators placed under location coordinates, left from the map on the main window.
Because of placement they are only visible if big_map option is turned on.

image

Please leave comments if this code needs corrections to be accepted.

### CODE REVIEW

**Comment by jezek** [comment]
I see, there is new code added to server, which sends new packets when drawing to client. Does the new server code works (have you tested it) with previous clients?

**Comment by EvgenijK** [comment]
> I see, there is new code added to server, which sends new packets when drawing to client. Does the new server code works (have you tested it) with previous clients?

Thanks for pointing on this. I tested and it didn't work.

I updated game version and added client version check in server functions.

**Comment by mhirki** [comment]
I think adding 7 new packet types is a total waste. Everything should be compressed into a single new packet with 32 bits reserved for flags.

**Comment by EvgenijK** [comment]
> I think adding 7 new packet types is a total waste. Everything should be compressed into a single new packet with 32 bits reserved for flags.

I see two variants of this:
1) Send flags to toggle the indicators, i.e. if the indicator flag is in the packet, then toggle it on interface. This way requires to store flags on client.
2)  Send flags as state of indicators - if flag is set in the packet, then draw indicator, if flag isn't set, then erase it. There is no need to store info in this variant, but we have to send info about all indicators each time.

I hope i explained this variants clearly enough.
Personally I more like second variant.

What do you think, what would be better solution?

**Comment by jezek** [comment]
@EvgenijK 
> What do you think, what would be better solution?

I would opt for the second variant too.

**Comment by mhirki** [comment]
Yeah, variant 2 is defintely the answer.

**Comment by EvgenijK** [comment]
@jezek @mhirki 
I made changes, waiting for the review.

File: src/server/xtra2.c [hunk 1]
```diff
@@ -616,6 +616,9 @@ bool set_tim_esp(int Ind, int v) {
 	/* Recalculate boni */
 	p_ptr->update |= (PU_BONUS | PU_MONSTERS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Comment (jezek) [inline src/server/xtra2.c:620]: Indentation mismatch?

File: src/server/xtra2.c [hunk 2]
```diff
@@ -2772,6 +2775,9 @@ bool set_oppose_acid(int Ind, int v) {
 	/* Recalculate boni */
 	p_ptr->update |= (PU_BONUS);
 
+	/* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Comment (jezek) [inline src/server/xtra2.c:2779]: Indentation mismatch?

File: src/server/xtra2.c [hunk 3]
```diff
@@ -2818,7 +2824,10 @@ bool set_oppose_elec(int Ind, int v) {
 	/* Recalculate boni */
 	p_ptr->update |= (PU_BONUS);
 
-	/* Handle stuff */
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Comment (jezek) [inline src/server/xtra2.c:2828]: Indentation mismatch?

File: src/server/xtra2.c [hunk 4]
```diff
@@ -2864,6 +2873,9 @@ bool set_oppose_fire(int Ind, int v) {
 	/* Recalculate boni */
 	p_ptr->update |= (PU_BONUS);
 
+    /* Redraw indicator */
+    p_ptr->redraw2 |= (PR2_INDICATORS);
```
Comment (jezek) [inline src/server/xtra2.c:2877]: Indentation mismatch?

File: src/server/xtra2.c [hunk 5]
```diff
@@ -2910,6 +2922,9 @@ bool set_oppose_cold(int Ind, int v) {
...
```
Comment (jezek) [inline src/server/xtra2.c:2926]: Indentation mismatch?

File: src/server/xtra2.c [hunk 6]
```diff
@@ -2956,6 +2971,9 @@ bool set_oppose_pois(int Ind, int v) {
...
```
Comment (jezek) [inline src/server/xtra2.c:2975]: Indentation mismatch?

File: src/server/xtra2.c [hunk 7]
```diff
@@ -3622,6 +3640,9 @@ bool do_divine_xtra_res(int Ind, int v) {
...
```
Comment (jezek) [inline src/server/xtra2.c:3644]: Indentation mismatch?

File: src/server/xtra1.c [hunk 8]
```diff
@@ -7327,6 +7366,11 @@ void redraw2_stuff(int Ind) {
 		p_ptr->redraw2 &= ~(PR2_MAP_SCR);
 		prt_map(Ind, TRUE);
 	}
+
+    if (p_ptr->redraw2 & PR2_INDICATORS) {
```
Comment (jezek) [inline src/server/xtra1.c:7370]: Indentation mismatch?

File: src/client/c-xtra1.c [hunk 9]
```diff
@@ -1027,6 +1027,128 @@ void prt_bpr(byte bpr, byte attr) {
 	Term_gotoxy(x, y);
 }
 
+void prt_indicators(u32b indicators) {
+    prt_res_fire((indicators & IND_RES_FIRE) != 0);
```
Comment (jezek) [inline src/client/c-xtra1.c:1031]: Indentation mismatch?

File: src/client/nclient.c [hunk 10]
```diff
@@ -177,6 +177,8 @@ static void Receive_init(void) {
 	receive_tbl[PKT_POWERS_INFO]	= Receive_powers_info;
 
 	receive_tbl[PKT_GUIDE]		= Receive_Guide;
+
+    receive_tbl[PKT_INDICATORS] = Receive_indicators;
```
Comment (jezek) [inline src/client/nclient.c:181]: Indentation mismatch?

File: src/client/nclient.c [hunk 11]
```diff
@@ -4840,8 +4842,19 @@ int Receive_Guide(void) {
 	return 1;
 }
 
+int Receive_indicators(void) {
+    int	n;
```
Comment (jezek) [inline src/client/nclient.c:4846]: Indentation mismatch?

File: src/common/defines.h [hunk 12]
```diff
@@ -2206,6 +2206,28 @@
...
+#define ROW_RESIST_ACID		23
+#define COL_RESIST_ACID		4
+
+#define ROW_RESIST_POIS		23
+#define COL_RESIST_POIS		5
+
+#define ROW_RESIST_MANA		23
+#define COL_RESIST_MANA		6
+
+#define ROW_TEMP_ESP		25
```
Comment (jezek) [inline src/common/defines.h:2227]: Indentation consistency?

File: src/server/nserver.c [hunk 13]
```diff
@@ -8850,6 +8850,19 @@ int Send_Guide(int Ind, byte search_type, int lineno, char* search_string) {
 	else return Packet_printf(&connp->c, "%c%c%d%s", PKT_GUIDE, search_type, lineno, search_string);
 }
 
+int Send_indicators(int Ind, u32b indicators) {
+    connection_t *connp = Conn[Players[Ind]->conn];
```
Comment (jezek) [inline src/server/nserver.c:8854]: Indentation mismatch?

File: src/server/xtra1.c [hunk 14]
```diff
@@ -684,6 +684,42 @@ static void prt_extra_status(int Ind) {
 	Send_extra_status(Ind, status);
 }
 
+static void prt_indicators(int Ind) {
+    player_type *p_ptr = Players[Ind];
```
Comment (jezek) [inline src/server/xtra1.c:688]: Indentation mismatch?

File: src/common/pack.h [hunk 15]
```diff
@@ -253,6 +253,7 @@
 
 #define PKT_GUIDE		196	/* Server issues Guide-invocation on client-side */
 
+#define PKT_INDICATORS 197
```
Comment (jezek) [inline src/common/pack.h:256]: I know, there are many misaligned constants, but try to align as the one before.

**Review by jezek — CHANGES_REQUESTED** [review]
Apart from some indentation consistency mismatch, I see no logical flaws.
Note: I haven't tested if it compiles or behaves as expected (yet).

File: src/server/xtra2.c (hunk 1 above)
Comment (EvgenijK) [reply src/server/xtra2.c]: fixed

File: src/server/xtra2.c (hunk 2 above)
Comment (EvgenijK) [reply src/server/xtra2.c]: fixed

File: src/server/xtra2.c (hunk 3 above)
Comment (EvgenijK) [reply src/server/xtra2.c]: fixed

File: src/server/xtra2.c (hunk 4 above)
Comment (EvgenijK) [reply src/server/xtra2.c]: fixed

File: src/server/xtra2.c (hunk 5 above)
Comment (EvgenijK) [reply src/server/xtra2.c]: fixed

File: src/client/nclient.c (hunk 11 above)
Comment (EvgenijK) [reply src/client/nclient.c]: fixed

File: src/server/nserver.c (hunk 13 above)
Comment (EvgenijK) [reply src/server/nserver.c]: fixed

File: src/server/xtra1.c (hunk 14 above)
Comment (EvgenijK) [reply src/server/xtra1.c]: fixed

File: src/common/pack.h (hunk 15 above)
Comment (EvgenijK) [reply src/common/pack.h]: fixed

File: src/common/defines.h (hunk 12 above)
Comment (EvgenijK) [reply src/common/defines.h:2227]: Tried to fixe this in CLion, Vim and github online editor, all of them shows that `25` on the same columt with other values

Comment (jezek) [reply src/common/defines.h:2227]: That's wierd, but you're right. When I look at the [whole file](https://github.com/TomenetGame/tomenet/blob/e8fcb013e0297dad3f39141cddeb686559e299f3/src/common/defines.h#L2224-L2227), the indetation is ok.

File: src/server/nserver.c [hunk 16]
```diff
@@ -8850,6 +8850,19 @@ int Send_Guide(int Ind, byte search_type, int lineno, char* search_string) {
...
 }
 
+int Send_indicators(int Ind, u32b indicators) {
+	connection_t *connp = Conn[Players[Ind]->conn];
+	//player_type *p_ptr = Players[Ind];
+
+	if (!BIT(connp->state, CONN_PLAYING | CONN_READY)) {
+		errno = 0;
+		plog(format("Connection not ready for indicators (%d.%d.%d)",
+					Ind, connp->state, connp->id));
```
Comment (jezek) [inline src/server/nserver.c:8860]: I would make this a oneliner. Other lines above are longer, so it wouldn't stand out and wrapping can do an editor.

**Review by jezek — CHANGES_REQUESTED** [review]
Thank you for indentation change. Only the 2 line logging caught my eye. Sorry for missing it the first time and is just a suggestion of me. The maintainers have the last word, so let's wait for their comments. When I have more free time, I will try to compile (server, client) and test it,

**Comment by jezek** [comment]
I've tested this PR. After merging to current mater, it compiles (server+client) on ubuntu 20.10 without introducing new warnings. I've also tested new (4.7.3.1) server against new (4.7.3.1) and old (4.7.3) client and vice versa. Everything works as described in this PR, no bugs spotted.

**Comment by mhirki** [comment]
Thanks for the testing jezek. I did my own testing and everything seems to work. I'm gonna merge the code now.

**Pull request merged**